### PR TITLE
DEPR: remove FrozenNDarray

### PR DESCRIPTION
**Pull request by jbrockmendel**
Reboots #29335.

### CODE REVIEW

File: doc/source/whatsnew/v1.0.0.rst [hunk 1]
```diff
@@ -433,6 +433,7 @@ or ``matplotlib.Axes.plot``. See :ref:`plotting.formatters` for more.
 - Removed previously deprecated :func:`pandas.tseries.plotting.tsplot` (:issue:`18627`)
 - Removed the previously deprecated ``reduce`` and ``broadcast`` arguments from :meth:`DataFrame.apply` (:issue:`18577`)
 - Removed the previously deprecated ``assert_raises_regex`` function in ``pandas.util.testing`` (:issue:`29174`)
+ - Removed the previously deprecated ``FrozenNDArray`` class in ``pandas.core.indexes.frozen`` (:issue:`29335`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: extra space

File: pandas/core/indexes/multi.py [hunk 2]
```diff
@@ -810,6 +811,13 @@ def set_levels(self, levels, level=None, inplace=False, verify_integrity=True):
     def codes(self):
         return self._codes
 
+    def _coerce(self, array_like, categories, copy=False):
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: can you make this a free function, add  types & a doc-string

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by WillAyd — APPROVED** [review]
nice lgtm. Thanks for picking this up!

File: pandas/core/indexes/multi.py [hunk 3]
```diff
@@ -3433,3 +3433,26 @@ def maybe_droplevels(index, key):
             pass
 
     return index
+
+
+def _coerce_indexer_frozen(array_like, categories, copy: bool = False) -> np.ndarray:
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: i would move this to pandas.core.dtypes.cast

**Review by jreback — CHANGES_REQUESTED** [review]
small request, rebase (i know need to rebase almost all.....)

File: pandas/core/indexes/multi.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/indexes/multi.py]: its only ever used here, and doesnt really have anything to do with dtypes

**Comment by jbrockmendel** [comment]
rebased+green

File: pandas/core/indexes/multi.py [hunk 4]
```diff
@@ -1095,7 +1094,8 @@ def _format_native_types(self, na_rep="nan", **kwargs):
             if mask.any():
                 nan_index = len(level)
                 level = np.append(level, na_rep)
-                level_codes = level_codes.values()
+                assert not level_codes.flags.writeable  # i.e. copy is needed
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: shoulnt' you be calling _coerce_indexer_frozen here (with copy=True)?

Comment (jbrockmendel) [reply pandas/core/indexes/multi.py]: no, _coerce_indexer_frozen returns an ndarrary that is not writeable, whereas here we have a non-writeable ndarray and need to get a writeable copy

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**